### PR TITLE
Isolating issues in handling rank-deficiency

### PR DESCRIPTION
**Pull request by palday**
The Cholesky decomposition can fail and thus far it fails silently, which leads to the sporadic errors we get in testing against Julia nightly. 

Possible solutions:
- Use SVD or QR decompositions to compute pivot
- Use SVD for more robust rank detection
- ????

This initial pull is just proof-of-concept for where the Cholesky fails on the CIs. No solution is implemented yet.

### CODE REVIEW

**Comment by palday** [comment]
Just had Cholesky pass on the nightly CI with this versioninfo:
```
 Julia Version 1.6.0-DEV.85
Commit 0413ef0e4d (2020-05-24 02:53 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU E5-2673 v4 @ 2.30GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, broadwell)
```

**Comment by palday** [comment]
And now from a failed run:
```
Julia Version 1.6.0-DEV.85
Commit 0413ef0e4d (2020-05-24 02:53 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Platinum 8171M CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
```

**Comment by palday** [comment]
So it really does look like the interaction of Intel Skylake and development version of Julia that's causing this problem. 

@andreasnoack Has anything changed in `LinearAlgebra` since 1.4 that would make the linear algebra more sensitive to processor details?

**Comment by palday** [comment]
Skylake processors do pass on Julia 1.4 / the older LLVM:
```
Julia Version 1.4.1
Commit 381693d3df* (2020-04-14 17:20 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Platinum 8171M CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-8.0.1 (ORCJIT, skylake)
```

**Comment by palday** [comment]
So it's not just the newer LLVM version on Skylake. I've found a non Xeon Skylake to run the tests on and they don't fail (OpenSuse 15.0):

```
Julia Version 1.6.0-DEV.90                                                                                            
Commit c832e47a0a (2020-05-25 08:56 UTC)      
Platform Info:                                                                                                        
  OS: Linux (x86_64-pc-linux-gnu)                                                                                     
  CPU: Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz                                                                       
  WORD_SIZE: 64                                                                                                       
  LIBM: libopenlibm                                                                                                   
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)                                                                               
Environment:                                                                                                          
  JULIA_LOAD_PATH = @:/tmp/jl_Py0VU3  
```

But even on the same nightly, they do fail on Skylake Xeon (Ubuntu 18.04):
```
Julia Version 1.6.0-DEV.90
Commit c832e47a0a (2020-05-25 08:56 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Platinum 8171M CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
Environment:
  JULIA_LOAD_PATH = @:/tmp/jl_6QaBGU
```

Meanwhile the broadwell CI machines continue to pass.

**Comment by andreasnoack** [comment]
> @andreasnoack Has anything changed in LinearAlgebra since 1.4 that would make the linear algebra more sensitive to processor details?

It might be the version of OpenBLAS. We upgraded recently and it could probably affect AVX512 systems. Try comparing
```julia
julia> BLAS.openblas_get_config()
"OpenBLAS 0.3.5  USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=32"
```

There have been issues with AVX512 kernels in OpenBLAS in the past but it might also be that you need to use a different tolerance when determining the rank.

**Comment by palday** [comment]
Skylake machine where the tests passed:
```
"OpenBLAS 0.3.9  USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=32"
```
Skylake where the tests failed:
```
OpenBLAS 0.3.9  USE64BITINT DYNAMIC_ARCH NO_AFFINITY SkylakeX MAX_THREADS=32
```

@andreasnoack It does look the OpenBLAS version is playing a role here.

**Comment by andreasnoack** [comment]
You can try starting julia with `OPENBLAS_CORETYPE=haswell julia` on the `SkylakeX` machine to verify that restricting OpenBLAS to the Haswell kernels actually fixes the issue. But again, it might just be because of the differences in rounding if the tolerance is too strict.

**Comment by palday** [comment]
Skylake but with Haswell BLAS on the problematic machine works fine.

```
Julia Version 1.6.0-DEV.90
Commit c832e47a0a (2020-05-25 08:56 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) Platinum 8171M CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
Environment:
  JULIA_LOAD_PATH = @:/tmp/jl_xDogEn

OpenBLAS 0.3.9  USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell MAX_THREADS=32
```

Going to try MKL on a different Skylake machine now.

**Comment by palday** [comment]
Given that the MKL also fails on that example on Skylake, I don't think this is OpenBLAS specific, but might be something with BLAS optimized for Skylake kernel / use of AVX512 instructions. I don't think the matrix in question is ill-conditioned, but I also don't know enough about how this interacts with AVX512 and at what level the issue arises / where to tune tolerances. For our tests, I can construct a simpler matrix to test the relevant functionality.  @andreasnoack Is there a relevant place where I should post this example for BLAS people tuning for AVX512?

**Comment by andreasnoack** [comment]
I've now taken a closer look at this PR and the failing test. When running on a machine with the SkylakeX kernels then the 43 pivot is small but positive whereas it's slightly negative when running with Haswell kernels. I think such differences are expected due to the rounding differences when using different float instructions.

**Comment by palday** [comment]
@andreasnoack Thanks!

**Comment by palday** [comment]
I think this is ready for review.

Actual changes:
1. Fix handling of tolerance in `statscholesky`.
2. Add in an extra rank check and complain when it doesn't match (need to benchmark this)
3. Modify rank deficiency tests to not fail on BLAS using new Intel SIMD instructions. 
4. Print version info and OpenBLAS info at the start of testing (useful debugging CIs).

Is (2) a good idea or should I just omit that?

File: src/linalg/statschol.jl [hunk 1]
```diff
@@ -6,17 +6,21 @@ retains the original order unless singularity is detected.  Columns that are
...
 right hand side in a left circular shift.
 """
-function statscholesky(xtx::Symmetric{T}, tol::Real = -1) where {T<:AbstractFloat}
+function statscholesky(xtx::Symmetric{T}, tol::Real = 0.0) where {T<:AbstractFloat}
     n = size(xtx, 2)
-    chpiv = cholesky(xtx, Val(true), tol = T(-1), check = false)
+    chpiv = cholesky(xtx, Val(true), tol = T(tol), check = false)
     chunp = cholesky(xtx, check = false)
-    r = chpiv.rank
+    r = rank(xtx) #chpiv.rank
```
[the remark this answers is not in the document]
Comment (palday) [reply src/linalg/statschol.jl:13]: No, not yet discussed and one of the things I'm worried about. I had originally introduced it as a debugging step and my preference would be to avoid using expensive computations here (even though this is only run once to construct the model) so that we can have very large fixed-effects matrices. There was some discussion on using a different decomposition to avoid any latent numerical stability issues with the Cholesky factorization.

**Comment by palday** [comment]
So I guess the thing to do is drop the eigenvalue-based rank estimate and change things to guarantee we're working with the largest pivots?

**Comment by andreasnoack** [comment]
I'm not sure what the right solution is. Using just the pivoted Cholesky with the default tolerance `(-1)` might be the best from a numerical perspective (when assuming that the eigenvalue decomposition is too expensive) but the "R" pivoting was chosen for a reason. @dmbates has explained to me (at least once) why it's useful with the "R" pivoting instead of normal pivoting but unfortunately I don't remember the argument.

**Comment by Nosferican** [comment]
I currently use
```
F = bunchkaufman(Hermitian(X̃' * Diagonal(wts) * X̃), true, check = false)
bkr = count(x -> abs(x) ≥ √eps(), F.D)
if bkr < size(F, 2)
  lin_ind = sort!(invperm(F.p)[1:bkr])
  X = X[:,lin_ind]
  F = bunchkaufman(Hermitian(X * Diagonal(wts) * X), true)
else
  lin_ind = collect(1:size(F, 2))
end
```
which is the default `factorize` method for `Hermitian`.
The next alternative in in terms of more robust not too expensive if through
```
F = qr(X .* wts)
qrr = count(x -> abs(x) ≥ √eps(), diag(F.R))
if qrr < size(F, 2)
  lin_ind = sort!(invperm(F.p)[1:qrr])
  X = X[:,lin_ind]
  F = qr(X .* wts), true)
else
  lin_ind = collect(1:size(F, 2))
end
```
`qr` being the default `factorize` method for dense rectangular matrices.

SVD does give some weird results sometimes and it is a pain to find good tolerances such that they play well with other software. If `bunchkaufman` doesn't solve it, then likely have to go `qr`.

**Comment by andreasnoack** [comment]
I think the "R pivoting" approach might have some issues. This is an extreme example but consider
```julia
julia> m, n, k = 100, 30, 20;

julia> X = [ones(m) (randn(m, k)*randn(k, n)).*exp10.(range(-6, stop=6, length=n))'];

julia> y = X*ones(n + 1) + 0.1*randn(m);

julia> myrank(F::LinearAlgebra.QRPivoted; tol=1e-8) = searchsortedlast(abs.(diag(F.R)), tol, rev=true);

julia> function myreg(X, y)
         F = qr(X, Val(true))
         r = myrank(F)
         coefs = F.Q'y
         coefs[(r+1):end] .= 0
         ldiv!(UpperTriangular(view(F.R, 1:r, 1:r)), view(coefs, 1:r))
         return coefs[invperm(F.p)]
       end;

julia> dR = rcopy(R"lm.fit($X, $y)");

julia> [dR[:coefficients] myreg(X, y)]
31×2 Array{Union{Missing, Float64},2}:
       1.00699     1.00699
      -1.85313e12  0.0
      -1.3466e12   0.0
      -7.89672e11  0.0
      -8.54843e10  0.0
      -3.71057e9   0.0
      -4.47597e9   0.0
       2.08636e10  0.0
      -2.4018e9    0.0
      -1.00694e8   0.0
      -1.17907e9   0.0
       1.47754e8   1.02591
       1.55519e8   0.954906
      -4.22778e6   0.888604
      -2.47268e7   1.05529
      -1.18547e7   1.01691
 -393723.0         0.989025
  637194.0         1.00004
  323120.0         1.00113
 -176288.0         0.999776
  -96793.1         1.00011
        missing    0.999929
        missing    1.0
        missing    1.00001
        missing    1.0
        missing    1.0
        missing    1.0
        missing    1.0
        missing    1.0
        missing    1.0
        missing    1.0
```
Notive that the rank estimates match
```julia
julia> myrank(qr(X, Val(true))) == dR[:rank]
true
```
It's not that the R solution is totally wrong since the prodictions in comparable
```julia
julia> norm(dR[:fitted_values] - X*myreg(X, y))
1.9030265205212798e-7
```
but I'm not sure how useful the R estimates are.

**Comment by dmbates** [comment]
I can describe the rationale for the pivoting approach in R but, before doing so, let me say that the rationale is now irrelevant.  Decisions in R are often "for compatibility with S" and the design decisions for S occurred in the 1970's and 1980's when computing equipment was, shall we say, less powerful than today.  (I had a blog called "A Second Megabyte of Memory" whose name came from my radical decision to by a second megabyte of memory for the Vax-11/750 computer that we bought for our department in 1983.  It came with 1 mb and I spend $7500 on a second mb when configuring it.)

If you have a full-rank n by p model matrix, X, and use the unpivoted QR decomposition then the first p elements of Q'y are used to solve for the coefficient estimates.  They also provide the sequential sums-of-squares for the model terms if you don't do any pivoting.  In S and later in R the formula language imposed an ordering on the terms in the models so that these sequential sums of squares were meaningful.  Also, the use of "contrasts" meant that for most cases the constructed model matrix was of full column rank.

In the unusual cases where X was rank deficient the idea was to move any columns that were dependent on columns to their left to the right hand side.  That way the term order and the interpretation of sequential sums of squares was preserved.

My objective here was to do something similar but that it probably unnecessary.  With a mixed-effects model you don't need to worry about rank deficiency in Z, the model matrix for the random effects, because of the regularization term in \Lambda'Z'Z\Lambda  + I.  In fact if you include the constant term then Z will always be rank deficient but you don't care.  So the only case you have to worry about is whether X is rank deficient and for that you could even use an SVD of X as UDV' if you wanted.  You can check the apparent rank and, if it is rank-deficient, then determine the dominant elements in V to see which columns in the original should be dropped and proceed.

It is probably better to check the rank using an SVD or a pivoted QR rather than the pivoted Cholesky or the Eigenvalue decomposition because of the inflation of the condition number when creating X'X.

**Comment by palday** [comment]
I had started tinkering with a pivoted QR decomposition, but I'll give both approaches a stab when I get a little bit of time / need a break from the other items on my MixedModels todo list.. My main priority was to find out what was causing the intermittent test failure and I've done that (and provided an alternative test that's quite as ill-conditioned).

**Comment by andreasnoack** [comment]
@dmbates Thanks for reminding me about the sequential sums of squares. I've heard it before but had completely forgotten. Since the reason to prefer that method doesn't apply here, I'd propose that you just use the pivoted Cholesky instead of `statscholesky`.

Using the Cholesky is so much faster than QR not to mention SVD so unless some cases show up where the squaring becomes a real problem, I think it would be the better choice. You'd probably still want to set `tol=-1` though. I think `tol=0.0` will make it identical to unpivoted Cholesky.

**Comment by palday** [comment]
I used `tol=0.0` to copy the `LinearAlgebra` documentation ... is that a change I should also make over there while doing the other documentation work I volunteered for?

**Comment by andreasnoack** [comment]
> is that a change I should also make over there while doing the other documentation work I volunteered for?

Hm. I'm in favor but we might not get away with it since it could be considered breaking. (Unfortunately for me, [I'm to blame for this](https://github.com/JuliaLang/julia/pull/5526) and I didn't even mention the change from `-1.0` to `0.0` in the PR. 😞 )

**Comment by palday** [comment]
I've done some limited benchmarking and functional examination of using QR on the model matrix or Cholesky on the crossproduct and the performance is comparable. Both yield the same permutation, which is actually the problem: the default permutation does things that are surprising statistically (at least coming from most other statistical software), like dropping lower-order terms (including the intercept) to preserve higher order terms. I need to think a bit whether there's a way to use the permutation returned from this decomposition to get one with our desired properties without using the somewhat numerically unstable iterative circular shift we currently have.

In other words, I once again realize how much of my life reduces to the first year of undergraduate maths.

**Comment by palday** [comment]
After a lot of back and forth ....

The old R behavior of shifting the columns to the end remains. *When it works*, it has the very desirable property that higher-order model terms are dropped before lower-order terms.

If we ever encounter a real-world example where this left circular shift fails (especially if it happens on multiple architectures), then I'm happy to revisit this. In the meantime, I've changed the test to one that passes on both Haswell and Skylake. Right now, this test failing is messing up our PR review process, so I'm willing to kick the can down the road. 

As to why the R behavior fails -- that's more interesting. The rank is given as `98`, but the `info` field has `100`, which, according to the [LAPACK documentation](https://www.netlib.org/lapack/explore-html/d1/d7a/group__double_p_ocomputational_ga2f55f604a6003d03b5cd4a0adcfb74d6.html), means that the 100th leading minor is not positive definite. But the first non positive definite leading minor should be less than the rank. We're at the edge of my LAPACK and BLAS knowledge, but this seems like weird behavior.

**Comment by dmbates** [comment]
@palday, @andreasnoack Bearing in mind that this check on the rank of the fixed-effects model matrix is only done once in the model construction phase, we could use something like a pivoted-QR to check the rank.  All we really want is a numerical rank, k, and a permutation such that the leading k columns are numerically independent.  I would be willing to sacrifice a rational ordering of the columns for reproducibility in the examples.

**Comment by palday** [comment]
There's a middle ground here that I can implement: we just wrap the BLAS pivoted Cholesky so that the order of the linearly independent columns isn't changed and the linearly dependent columns are put all the way to the right. In other words, we let BLAS decide which columns are redundant, but otherwise preserve order.

**Comment by palday** [comment]
I wrote that last comment and did some tests .... and it seems that the original test that started all of this out is so ill-conditioned that even the BLAS delivers different answers on different architectures as to which columns should be pivoted. But then again, so does the QR decomposition. In other words, swapping to QR won't save us here.

So then: we either have a really ill conditioned example (inclusive) or there is a latent issue with the use of AVX instructions in OpenBLAS.

**Comment by Nosferican** [comment]
Guess we should well define what the best behavior should be. Do we want to keep the full rank version based on the column ordering, do we want to make sure to always prefer the intercept, do we want to keep the combinations for a categorical variables over just a part set? Sometimes in applied work on wants to make sure to include the feature of interest, yes, but other than that and the intercept I am not sure. I really dislike Stata's random algo to choose which ones to drop so for it to be deterministic to some extent seems desirable. I would maybe use some measure of variability in the column as a way to "prioritize" the selected columns.

**Comment by andreasnoack** [comment]
So a couple of thoughts a had after rereading this issue:

1. Mixing rank determination strategies is not a good idea as they can easily disagree. Any solution should be based on a single procedure. (As I read the thread, this seems to be the consensus as well.)
2. The pivoted QR and pivoted Cholesky are equivalent in full precision so I believe it's expected that they give the same conclusion whenever you use the default tolerance but if you use a lower tolerance the QR should be able to be more precise than Cholesky. However, I don't think it would be useful with the extra precision in this application. (I might be wrong)
3. It's probably not possible to ensure that the rank determination is consistent across architectures in all cases. However, in should probably be the case in most real applications and it might, therefore, make sense to adjust the test if it's pathological.
4. Maybe you want to consider a higher tolerance than the default one. It should make is less likely that noise accidentally increases the rank with a large effect on coefficients. I guess it would also make the results less sensitive to architecture.
5. I'm wondering if it would be possible to get results that are easier to interpret when using the pivoted Cholesky if the variables are scaled differently. I think the is lost because higher-order terms end up with a larger column norm than the intercept but since the intercept doesn't have to be one, it's, in some sense, a self-created problem. Make the intercept `1e6` and it will have the largest norm and probably not get dropped. Hence, it might be possible to use a scaling that ensures lower order terms have a large norm and then apply the inverse scaling after the rank determination.

**Comment by palday** [comment]
@dmbates If you care to review, we can be done with this mess! The pivoted QR is now what we're using; I've left the pivoted Cholesky in for now and made it clear in the docs that we make no promise that we'll keep using the same algorithm for rank determination and pivoting.

 Do check out the docs -- they're not my best explanation ever, but if they're good enough and I haven't been infelicitous in my simplifications, we can merge this change in and improve the docs later. Then the CIs will all be happy again.

**Comment by dmbates** [comment]
I have made some suggested edits in `rankdeficiency.md`.  These are suggestions only.

One thing I did try to distinguish more clearly is the distinction between singularity of the covariance matrix estimates for the random effects, which will result in the conditional means being on some kind of hyperplane in one or more dimensions, and singularity or rank deficiency in the random-effects model matrix, which essentially always occurs.  The random-effects model matrix for any model with a random intercept contains a set of indicators for the levels of the grouping factor.  Even without considering `fulldummy` these columns sum to an intercept column which makes the concatenation of `X` and `Z` rank deficient.  So we always have some type of rank deficiency for the random-effects model matrix but it is not a problem because of the shrinkage or regularization.

**Comment by palday** [comment]
Thanks @dmbates! I think it reads better now and sharpening the distinction between the two types of singularity in the RE is a nice addition.

Also, I apparently cannot type `redudant.`

**Comment by palday** [comment]
If you're happy, then let's squash and merge (and please remove all the redundant stuff from the long-form commit message).

**Comment by dmbates** [comment]
Can you rebase? I tried to do so and it didn't go well.

**Comment by palday** [comment]
I can, but there's no need to do so: the squash will collapse everything into a single commit and you can edit the commit message through the web interface. Should I just do that?

**Comment by dmbates** [comment]
Yes, pl
